### DISCUSSION TITLE
Make the "embed" Pocketbase JS SDK available as a ES6 module from within the application.

### DISCUSSION THREAD

**Discussion by dschissler** [pocketbase/pocketbase#1135] [Q&A]
In regards to [embedded Javascript migration capabilities](https://github.com/pocketbase/pocketbase/commit/328b99a690275232a913a00bc2ef1bc2981d7286#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4edR9-R12)

Since you already need to embed the Pocketbase Javascript SDK for the admin panel then perhaps that could be available within the entire application as a compiled ES module that exports the necessary stuff.  I mostly don't use typescript and so I personally wouldn't miss the type defs.  Others would and so they could simply choose not to use that.

The data is already taking up space in the binary, so might as well make it available for those who want to use it.  For those who do use this method then they'd always be up on the most compatible version of the kit.

My thinking is that if you refactored how this Javascript code is embedded then you could make it easy to understand how to import it both client side as well as server side (for the migrations).

Also I want to point out that having the Pocketbase SDK provided as a separate bundle / ES6 module would make an extra request to the Pocketbase server since it wouldn't just all be bundled up as one.  That seems to me to be inconsequential for the admin panel.  Then for those who either need the Typescript defs or need extreme efficiency then they can just do it how they're doing it now.

Another benefit to having Pocketbase embedded as a ES module is that if could make tutorials, examples and first use easier because they wouldn't need to have a Node and web bundle setup for their front end already.  That makes it much easier to get up and running testing out the viability of the whole suite.  Just create a static html page and do the work inline inside of a script tag.  During this time they'll be screwing around in the admin panel, querying records and seeing how it works.  Then finally comes the very time consuming bundler setup.

**Comment by ganigeorgiev** [comment]
I think there is some confusion. The embedded JS interpreter that allows running JS migrations (and eventually to bind to event hooks) is not related to the client-side JS SDK. The client-side SDK works only with the PocketBase web APIs, where the linked changes execute directly server-side as if they were part of the Go code.

I don't think exposing the JS SDK will be very useful. Most developers nowadays use bundlers (rollup, webpack, etc.) and are accustomed to load external JS code using `import/require` statements.
If they need to load the SDK as static url, they can always load the dist js file manually or use some CDN assets service, eg.:
`https://cdn.jsdelivr.net/gh/pocketbase/js-sdk@master/dist/pocketbase.umd.js`.

**Reply by dschissler** [reply]
I understand that the two are different techs.  It's just that you're already Golang embedding Pocketbase into the binary and that comes to 35KB.  Then we're embedding it again when we use PB as a framework.

> If they need to load the SDK as static url, they can always load the dist js file manually or use some CDN assets service, eg.:
https://cdn.jsdelivr.net/gh/pocketbase/js-sdk@master/dist/pocketbase.umd.js.

That makes it impossible to use with a locked down HSTS.

**Reply by ganigeorgiev** [reply]
I'm sorry, I don't understand your use case. Could you elaborate a little more? The external CDN was just an example, you can always load it manually by downloading the bundle from the `dist` directory - https://github.com/pocketbase/js-sdk/tree/master/dist.

Also note that the PocketBase version is not tied to the JS SDK version (aka. for example you can have PocketBase v0.9.0 backend and JS SDK v0.8.1).

**Reply by dschissler** [reply]
My use case is just that the SDK is already being embedded into the binary and framework (both ways) but it's not directly reachable because it's bundled into the admin UI.  I'm thinking that it would be useful for a low-tech **and self-contained** Pocketbase build if node.js wasn't required to use Pocketbase as a framework.  All that needs to happen is that the Pocketbase SDK is saved / go embedded in its own ES module file (basically how it is in the NPM package).

Don't get me wrong.  I use node with Webpack, Rollup and Vite and Svelte, but it's all super complicated.  I think that it might appeal to some people sometimes to be able to just drop in JS files to get up and running without that whole multi-year learning curve, or copying over a complex project skeleton.

Pocketbase is cool and fresh because you can get a very functional database without worrying about setting it up.  The ACME autocert feature removes a lot of crap for beginners and I think that some interesting things could happen if it also wasn't strictly necessary to have to be a web bundling wizard to have local access to what is already there (in an unusable form).

Also there are some newer JS frameworks which compile cleanly in the browser.  That is basically the new jQuery without a bundler approach that can get a lot done at the cost of startup and efficiency.  Also, without needing Node then a build tool becomes less a lot more possible.  This allows rapid prototyping of a single thing in a new server.

I'd like to be able to use Pocketbase as a framework for rapid prototyping with just `go run program.go`.  No build tool.  No bundler.  No Node.

**Reply by ganigeorgiev** [reply]
@dschissler I think understand what do you want, but I still don't think this will be very useful because the users will be "locked" to the SDK version that the Admin UI is using and as mentioned previously the PocketBase backend version and the JS SDK versions are not tied up and it is possible that a bugfix/feature in the JS SDK to not require new PocketBase release. Additionally, if we export the es module from the admin ui, this will require documenting it and explaining that that the SDK version cannot be changed, which in my opinion will be confusing.

If users want to run the JS SDK without a package manager and/or bundler, they have 2 options:

1. download the dist file locally (UMD or ES depending on the usage; this is mentioned in the JS SDK ["Browser (manually via script tag)"](https://github.com/pocketbase/js-sdk#browser-manually-via-script-tag) instructions)

2. use an external CDN that will do 1) for us:
  https://cdn.jsdelivr.net/gh/pocketbase/js-sdk@v0.8.1/dist/pocketbase.umd.js

**Reply by dschissler** [reply]
What about a single checkbox in the Admin panel that allows a static public directory to take over all routes (minus the ones defined by Pocketbase)?  That would make it so that someone could start playing with Pocketbase right away.  For example, very admin stuff or prototyping.

**Reply by ganigeorgiev** [reply]
> What about a single checkbox in the Admin panel that allows a static public directory to take over all routes

I think we already do this with the `pb_public` directory or do you mean something else?

**Reply by dschissler** [reply]
Cool.  Somehow I missed that.
